flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/EnforceLocalSortAggRule.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.rules.physical.batch
+
+import org.apache.flink.table.planner.plan.nodes.FlinkConventions
+import org.apache.flink.table.planner.plan.nodes.physical.batch.{BatchExecExchange, BatchExecExpand, BatchExecSort, BatchExecSortAggregate}
+
+import org.apache.calcite.plan.RelOptRule.{any, operand}
+import org.apache.calcite.plan.RelOptRuleCall
+import org.apache.calcite.rel.{RelCollationTraitDef, RelNode}
+
+/**
+  * An [[EnforceLocalAggRuleBase]] that matches [[BatchExecSortAggregate]]
+  *
+  * for example: select count(*) from t group by rollup (a, b)
+  * The physical plan
+  *
+  *  {{{
+  * SortAggregate(isMerge=[false], groupBy=[a, b, $e], select=[a, b, $e, COUNT(*)])
+  * +- Sort(orderBy=[a ASC, c ASC, $e ASC])
+  *    +- Exchange(distribution=[hash[a, b, $e]])
+  *       +- Expand(projects=[{a=[$0], b=[$1], $e=[0]},
+  *                           {a=[$0], b=[null], $e=[1]},
+  *                           {a=[null], b=[null], $e=[3]}])
+  * }}}
+  *
+  * will be rewritten to
+  *
+  * {{{
+  * SortAggregate(isMerge=[true], groupBy=[a, b, $e], select=[a, b, $e, Final_COUNT(count1$0)])
+  * +- Sort(orderBy=[a ASC, c ASC, $e ASC])
+  *    +- Exchange(distribution=[hash[a, b, $e]])
+  *       +- LocalSortAggregate(groupBy=[a, b, $e], select=[a, b, $e, Partial_COUNT(*) AS count1$0]
+  *          +- Sort(orderBy=[a ASC, c ASC, $e ASC])
+  *             +- Expand(projects=[{a=[$0], b=[$1], $e=[0]},
+  *                                 {a=[$0], b=[null], $e=[1]},
+  *                                 {a=[null], b=[null], $e=[3]}])
+  * }}}
+  */
+class EnforceLocalSortAggRule extends EnforceLocalAggRuleBase(
```

Comment:
I'm curious about this, when this is useful?

Reply:
just like `HashAggregate`,  it's useful when the planner choose `SortAggregate`. `LocalSortAggregate` will also reduce shuffling data and the processing data  for the global `SortAggregate` corresponding to `a=[null], b=[null], $e=[3]`.

Reply:
Actually it's because sometimes we can't do hash agg, like when udaf is involved.